### PR TITLE
Add strategy autogenerate the proxy class when the proxy file does no…

### LINKED ISSUE #362: Warning with php rename function, in ProxyGenerator class
Hello,

Sometimes i got a warning with the `rename` function, called in `Doctrine\Common\Proxy\ProxyGenerator:305`

```
Warning: rename(C:\Windows\TEMP\__CG__AppEntitiesCategory.php.551eee3a298629.89032411,C:\Windows\TEMP\__CG__AppEntitiesCategory.php)
```

I got this warning when I send multiple xhr request, so i guess there is a conflict with temporaries files ?

I temporarily fixed this by adding a `@` before the function call to disable error, but i hope there's a better way to fix this

Note: I got the same warning when I configure a specific cache directory

```php
// this directory exists and is writeable
Setup::createConfiguration($isDevMode, __DIR__.'/../../cache/doctrine');
```

### PR DESCRIPTION
**Pull request by Koc**
…t exist or outdated.

This should improve DX a lot. Fixes #362, #327

### CODE REVIEW

File: lib/Doctrine/Common/Proxy/AbstractProxyFactory.php [hunk 1]
```diff
@@ -216,6 +226,13 @@ private function getProxyDefinition($className)
                     require $fileName;
                     break;
 
+                case self::AUTOGENERATE_FILE_OUTDATED:
+                    if ( ! file_exists($fileName) || filemtime($classMetadata->getReflectionClass()->getFileName()) > filemtime($fileName)) {
```
Comment (Ocramius) [inline lib/Doctrine/Common/Proxy/AbstractProxyFactory.php]: what if the original class file name doesn't exist?

Comment (Koc) [inline lib/Doctrine/Common/Proxy/AbstractProxyFactory.php]: I haven't know that proxy generation can be used for internal classes. In this case we shouldn't generate new proxy class. What do you think?

Comment (Ocramius) [inline lib/Doctrine/Common/Proxy/AbstractProxyFactory.php]: It should probably re-generate it

**Comment by johndodev** [comment]
+1

**Comment by Koc** [comment]
@Ocramius I've updated the PR. IMHO there is no sense autogenerate proxy class for internall classes on each request because they are changes only when PHP/extension version changes, but in this case cache should be flushed anyway.

**Comment by Ocramius** [comment]
Checking this, it's not a sufficient change. Specifically, we'd need to track all ancestors in a way similar to https://github.com/doctrine/annotations/pull/105

A simpler approach would be to use something like https://github.com/Ocramius/PackageVersions to simply lock dependencies at package level or at entire project level, but it would require us to code-sign our proxies.

**Comment by Koc** [comment]
good catch, thanx for giving right direction

**Comment by mikeSimonson** [comment]
@Koc Should I close this PR and https://github.com/doctrine/DoctrineBundle/pull/636 then ?

**Comment by Koc** [comment]
Please no. I will continue on it. This problem occurs very ofthen on concurrent requests in dev env

**Comment by Majkl578** [comment]
Hello, I'm going to close this PR as we will not be maintaining Doctrine\Common\Proxy anymore. As an alternative, you can use [Ocramius/ProxyManager. ](https://github.com/ocramius/proxymanager). Thank you for the PR.
If you think this PR is absolutely must, let me know, I'll reopen and we can decide.

**Pull request closed without merging**